networking/src/Ntp/Client.hs

```diff
@@ -167,15 +172,22 @@ sendLoop cli addrs = do
             -- wait for responses and update status
             _ <- timeout respTimeout waitForResponses
             updateStatus cli
+            -- after @'updateStatus'@ @'ntpStatus'@ is guaranteed to be
+            -- different from @'NtpSyncPending'@, now we can wait until it was
+            -- changed back to @'NtpSyncPending'@ to force a request.
+            waitForRequest
         )
         (\a -> do
             -- send packets and wait until end of poll delay
             sock <- atomically $ readTVar $ ncSockets cli
             pack <- mkNtpPacket
             sendPacket sock pack addrs
 
-            threadDelay $ fromIntegral poll
-            cancel a
+            cmd <- timeout poll (wait a)
+            case cmd of
+                Nothing -> cancel a
+                Just SendRequest
+                        -> sendLoop cli addrs
```

Comment:
I don't think that's a real danger, it can only be triggered by the end user action.  But maybe there is a simple solution to get tail recursion here, I'll look into that.

Reply:
If we recurse here, is it still necessary to do lines 194-196? If not, put them in the `Nothing` case.

Reply:
I think you're right, we don't need to recurs here at all. It's fine if we just `return ()` and recurs only in the tail position. The `timeout poll (wait a)` will wait until either the `poll` timeout or somebody (re)set the `ncStatus :: TVar NtpStatus`.